pypeit/datamodel.py

```diff
@@ -0,0 +1,942 @@
+                `astropy.io.fits.ImageHDU`_ or
+                `astropy.io.fits.BinTableHDU`_). If true, the method
+                constructs an `astropy.io.fits.HDUList` with a
+                primary HDU, such that this call::
+
+                    hdu = fits.HDUList([fits.PrimaryHDU()] + self.to_hdu())
+
+                and this call::
+
+                    hdu = self.to_hdu(add_primary=True)
+
+                are identical.
+
+        Returns:
+            :obj:`list`, `astropy.io.fits.HDUList`_: A list of HDUs,
+            where the type depends on the value of ``add_primary``.
+        """
+        # Bundle the data
+        data = self._bundle()
+
+        # Construct the list of HDUs
+        hdu = []
+        for d in data:
+            if isinstance(d, dict) and len(d) == 1:
+                ext = list(d.keys())[0]
+                hdu += [io.write_to_hdu(d[ext], name=ext)]
+            else:
+                hdu += [io.write_to_hdu(d)]
+        return fits.HDUList([fits.PrimaryHDU()] + hdu) if add_primary else hdu
+
+    @classmethod
+    def from_hdu(cls, hdu):
+        """
+        Instantiate the object from an HDU extension.
+
+        This is primarily a wrapper for :func:`_parse`.
+
+        Args:
+            hdu (`astropy.io.fits.HDUList`_, `astropy.io.fits.ImageHDU`_, `astropy.io.fits.BinTableHDU`_):
+                The HDU(s) with the data to use for instantiation.
+        """
+        # NOTE: We can't use `cls(cls._parse(hdu))` here because this
+        # will call the `__init__` method of the derived class and we
+        # need to use the `__init__` of the base class instead. So
+        # below, I get an empty instance of the derived class using
+        # `__new__`, call the parent `__init__`, and then return the
+        # result. This is the first time I've used `__new__` so we may
+        # want to tread carefully, but it seems to work.
+        self = super().__new__(cls)
+        super(cls, self).__init__(cls._parse(hdu))
+        return self
+
+    def to_file(self, ofile, overwrite=False, checksum=True):
+        """
+        Write the data to a file.
+
+        This is a convenience wrapper for :func:`to_hdu` and
+        :func:`pypeit.io.write_to_fits`. The output is always placed
+        in the 2nd extension; the first (primary) extension is always
+        empty.
+
+        Args:
+            ofile (:obj:`str`):
+                Fits file for the data. File names with '.gz'
+                extensions will be gzipped; see
+                :func:`pypeit.io.write_to_fits`.
+            overwrite (:obj:`bool`, optional):
+                Flag to overwrite any existing file.
+            checksum (:obj:`bool`, optional):
+                Passed to `astropy.io.fits.HDUList.writeto`_ to add
+                the DATASUM and CHECKSUM keywords fits header(s).
+        """
+        io.write_to_fits(self.to_hdu(add_primary=True), ofile, overwrite=overwrite,
```

Comment:
If there is no hdr provided, can we use `pypeit.io.initialize_header()` as the default for Primary?

Reply:
Good idea!  Done.